MotionSensor, DoorSensor and PushButton

benfrancis opened the pull request:
These are all just specialisations of a BinarySensor with a single boolean property. I haven't defined any events yet. There are lots of events it's tempting to add, but I want to be sure they're definitely adding something over just property changes alone. We may need to add some events for the push button to detect certain time-sensitive interactions like double press or long press.

Can be viewed live at https://benfrancis.github.io/schemas/

Fixes #8
Fixes #9
Fixes #10

benfrancis commented:
These are all just specialisations of a BinarySensor with a single boolean property. I haven't defined any events yet. There are lots of events it's tempting to add, but I want to be sure they're definitely adding something over just property changes alone. We may need to add some events for the push button to detect certain time-sensitive interactions like double press or long press.

Can be viewed live at https://benfrancis.github.io/schemas/

benfrancis commented:
See also: #8, #9 and #10

rzr commented on index.html:
```diff
@@ -388,6 +394,72 @@ <h3>Events</h3>
+      </section>
+      <section id="PushButton">
+        <h2>PushButton</h2>
+        <p>A momenetary switch which can be pressed and returns to an un-pressed state when released.</p>
```
ok

https://github.com/mozilla-iot/schemas/issues/8

mrstegeman replied:
s/momenetary/momentary/

mrstegeman commented on index.html:
```diff
@@ -412,6 +484,42 @@ <h2>OnOffProperty</h2>
+            with the added semantics that <code>true</code> means "pushed" and
+            <code>false</code> means "not pushed".</p>
+      </section>
+      <section id="OpenClosedProperty">
```
Could this instead be `OpenProperty`? `OpenClosedProperty` by itself doesn't clearly indicate what `true` means.

rzr commented on index.html:
```diff
@@ -474,51 +474,51 @@ <h2>BooleanProperty</h2>
-            <code>false</code> means "no motion".</p>
       </section>
       <section id="PushedProperty">
         <h2>PushedProperty</h2>
```
Used past sense is weird here, I would have use "Pushable" but I am not a native English speaker.

rzr commented on index.html:
```diff
@@ -440,7 +440,7 @@ <h3>Events</h3>
       <section id="PushButton">
         <h2>PushButton</h2>
-        <p>A momenetary switch which can be pressed and returns to an un-pressed state when released.</p>
+        <p>A momentary switch which can be pushed and returns to an un-pushed state when released.</p>
```
isnt unpushed same as released ?

rzr replied:
IMHO,  "momentary switch" is adding confusion since switch are holding the state.. so button would be preferred as everyone know what it is. Eventually a movable button with a spring to keep it released if not pushed.

am I nitpicking  too much ?

benfrancis replied:
Yes un-pushed is the same as released. Writing "returns to an released state when released" would be a bit confusing. Adjective vs. verb again.

I'm not sure I understand your second comment. PushButton differs from OnOffSwitch in that it is a "momentary" or "biased" switch which returns to an un-pushed state when you remove your finger. These are generally known as a [push button](https://en.wikipedia.org/wiki/Push-button).

rzr replied:
my 2d comment was about the "A momentary switch" description of "PushButton", well nevermind it's ok to tell that a button is like a switch returning back to initial position if not hold.

benfrancis commented:
I've addressed the review comments.

There are lots of events I'm tempted to add:

PushButton:
- ClickEvent
- LongPressEvent
- PushedEvent
- ReleasedEvent

MotionSensor:
- MotionDetectedEvent

DoorSensor:
- OpenedEvent
- ClosedEvent

The thing is that none of these really add functionality which can't be achieved by just listening for property changes. ClickEvent and LongPressEvent are arguably time dependent and could therefore be useful.

Perhaps we should start with just the properties and see whether the events are actually needed?

benfrancis commented:
Related: I'm still waiting for an answer on https://github.com/iot-schema-collab/iotschema/issues/32
